src/core/domain-services/remote-key-service.ts

```diff
@@ -0,0 +1,196 @@
+        }
+        const methodData = this.generateMethodData(method, args);
+        console.log("RemoteKeyClient::Inside Invoke, RemoteUserId, MethodData", this.remoteUserId, methodData);
+        // @ts-ignore
+        await this.cruxProtocolMessenger.send({
+            content: methodData,
+            type: "KEY_MANAGER_REQUEST",
+        }, this.remoteUserId);
+        return methodData.invocationId;
+    }
+
+    public listenToInvocation = (invocationId: string, resultCallback: (msg: any, senderId: CruxId | undefined) => any, errorCallback: (err: any) => any): void => {
+        if (!this.cruxProtocolMessenger) {
+            throw Error("RemoteKeyClient cannot listen with no secureCruxNetwork");
+        }
+        console.log("RemoteKeyClient::ListenToInvocation::invocationId", invocationId);
+        this.emitter.on(invocationId, resultCallback);
+        this.emitter.on("error", errorCallback);
+    }
+
+    private generateMethodData(method: string, args: any[]) {
+        return {
+            args,
+            invocationId: makeUUID4(),
+            method,
+        };
+    }
+}
+
+export class RemoteKeyHost {
+    private cruxProtocolMessenger: CruxProtocolMessenger;
+    private keyManager: IKeyManager;
+
+    constructor(cruxProtocolMessenger: CruxProtocolMessenger, keyManager: IKeyManager) {
+        this.keyManager = keyManager;
+        this.cruxProtocolMessenger = cruxProtocolMessenger;
+    }
+
+    public async initialize() {
+        this.cruxProtocolMessenger.on("KEY_MANAGER_REQUEST", async (msg: any, senderId: CruxId | undefined) => {
+            console.log("Inside RemoteKeyHost::in::Msg, senderId: ", msg, senderId);
+            const data = await this.handleMessage(msg);
+            console.log("Inside RemoteKeyHost::initialize::Data(handleMessage): ", data);
+            this.sendInvocationResult(data, senderId!);
+        });
+    }
+    private async sendInvocationResult(result: any, receiverId: CruxId) {
+        if (!this.cruxProtocolMessenger) {
+            throw Error("RemoteKeyClient cannot send with no selfMessenger");
+        }
+        const resultData = this.generateInvocationResponse(result);
+        console.log("RemoteKeyHost::Inside sendInvocationResult::resultData: ", resultData);
+        // @ts-ignore
+        await this.cruxProtocolMessenger.send({
+            content: resultData,
+            type: "KEY_MANAGER_RESPONSE",
+        }, receiverId);
+    }
+
+    private async handleMessage(message: any) {
+        if (!VALID_METHODS.includes(message.method)) {
+            throw new Error("Invalid key manager method");
+        }
+        if (!this.keyManager) {
+            throw new Error("Key Manager not available");
+        }
+        let data;
+        if (message.method === "signWebToken") {
+            console.log("HandleMessage entrance:signWebToken");
+            data = await this.keyManager.signWebToken(message.args[0]);
```

Comment:
what will happen if function has multiple arguments? does args[0] mean only first argument will go?

Reply:
all functions require single or zero parameters. So args[0] will work

Reply:
Better way to do it which will work for any function - 

`            const func = (a, b) => {
                console.log(a+b);
            }
            const args = [1,2]
            // @ts-ignore
            func(...args);
`


or 

`            const func = (a, b) => {
                console.log(a+b);
            }
            const args = [1,2]
            func.apply(this, args);
`

Reply:
understood will modify it accordingly